staging/src/k8s.io/apiserver/pkg/util/flowcontrol/watch_tracker.go

```diff
@@ -51,37 +57,90 @@ type ForgetWatchFunc func()
 // of watches in the system for the purpose of estimating the
 // cost of incoming mutating requests.
 type WatchTracker interface {
-	// RegisterWatch reqisters a watch with the provided requestInfo
+	// RegisterWatch reqisters a watch based on the provided http.Request
 	// in the tracker. It returns the function that should be called
 	// to forget the watcher once it is finished.
-	RegisterWatch(requestInfo *request.RequestInfo) ForgetWatchFunc
+	RegisterWatch(r *http.Request) ForgetWatchFunc
 
 	// GetInterestedWatchCount returns the number of watches that are
 	// potentially interested in a request with a given RequestInfo
 	// for the purpose of estimating cost of that request.
 	GetInterestedWatchCount(requestInfo *request.RequestInfo) int
 }
 
+// builtinIndexes represents of set of indexes registered in
+// watchcache that are indexing watches and increase speed of
+// their processing.
+type builtinIndexes map[string]string
+
+func getBuiltinIndexes() builtinIndexes {
+	// The only existing indexes as of now are:
+	// - spec.nodeName for pods
+	// - metadata.Name for nodes, secrets and configmaps
+	// However, we can ignore the latter, because the requestInfo.Name
+	// is set for them (i.e. we already catch them correctly).
+	return map[string]string{
+		"pods": "spec.nodeName",
+	}
+}
+
 // watchTracker tracks the number of watches in the system for
 // the purpose of estimating the cost of incoming mutating requests.
 type watchTracker struct {
-	lock sync.Mutex
+	// indexes represents a set of registered indexes.
+	// It can't change after creation.
+	indexes builtinIndexes
 
+	lock sync.Mutex
 	watchCount map[watchIdentifier]int
 }
 
 func NewWatchTracker() WatchTracker {
 	return &watchTracker{
 		watchCount: make(map[watchIdentifier]int),
+		indexes:    getBuiltinIndexes(),
```

Comment:
I would order the fields in this construction the same as in the struct definition.

Reply:
done